### ISSUE TITLE
Consistency of controllable components [NDS-38] [NDS-39] [NDS-40] [NDS-41] [NDS-42] [NDS-43]

### ISSUE THREAD

**Issue by sh0ji** [wwnorton/design-system#107]
We need a more consistent approach to how our components with internal state can be used in a controlled and an uncontrolled way. In general, components are controlled by supplying it with a prop that controls some state (i.e., `checked`) and an event callback prop that triggers when that state is trying to change internally (i.e., `onChange`).

In other words, controlling a component requires two props to be controlled. Here is what I would propose:

- A **state** prop
  - When defined, the component's internal state should be ignored.
  - When `undefined`, the component's internal state should be used.
- A **default${State}** prop
  - When defined, this should set the component's initial internal state.
  - When `undefined`, the component should define its own initial internal state.
- An **event callback** prop
  - When defined, user interaction effects should be passed to the callback, having no effect on internal state.
  - When `undefined`, user interaction should change internal state.

|                          | State defined            | State undefined       |
| ------------------------ | ------------------------ | ----------------------- |
| **Callback defined**     | Controlled               | Controlled but doesn't change* |
| **Callback undefined** | Controlled but doesn't change* | Uncontrolled            |

\* The component should effectively not work if only one is defined.

@JRebella, does this match with your expectations?

[NDS-38] [NDS-39] [NDS-40] [NDS-41] [NDS-42] [NDS-43]

[NDS-38]: https://wwnorton.atlassian.net/browse/NDS-38
[NDS-39]: https://wwnorton.atlassian.net/browse/NDS-39
[NDS-40]: https://wwnorton.atlassian.net/browse/NDS-40
[NDS-41]: https://wwnorton.atlassian.net/browse/NDS-41
[NDS-42]: https://wwnorton.atlassian.net/browse/NDS-42
[NDS-43]: https://wwnorton.atlassian.net/browse/NDS-43

**Comment by sh0ji** [comment]
Below are all the components that meet these criteria in some way, and how they match or don't match this pattern. I'll update this list as I find more issues.

## Checkbox & Radio

Both of these components wrap Choice, which itself is an `<input type={radio/checkbox}>` under the hood. We're using the DOM interface for this (`HTMLInputElement`), which exposes the following two props:

* state: `checked`
* event: `onChange`
* **behaves as outlined in the proposal**

## Disclosure

Since our disclosure wraps [the `details` element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/details), the internal state that must be controlled is [the details' `open` attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/details#attr-open). Since our disclosure animates open/closed, we've broken the events into 6 callback props, only two of which affect the internal `open` state.

* state: `isOpen` (open/closed)
  * This is ignored after the first render, making it impossible to fully control this component. Note: this should be considered a bug.
* events: `onOpenStart` triggers when `isOpen` becomes `true` & `onCloseEnd` triggers when `isOpen` becomes `false`.
  * Returning `false` causes these callbacks to do nothing.
  * Returning a `Promise` will await that promise and then initiate the transition. This gives you time to update your state, but it is not then used to update the component.
* **does not behave as outlined in the proposal**

## Switch

* state: `checked` (on/off)
* event: `onToggle`.
  * This triggers when the internal state changes.
  * Unintuitively and perhaps incorrectly, the `onClick` event makes `checked` controllable. This should probably be fixed in a refactor.
  * Side note: we should probably deprecate this and rename it `onChange` for consistency.
* **does not behave as outlined in the proposal**

## Dropdown

This controls two internal pieces of state: `isOpen` (whether the list box is rendered), and `selected` (what option is selected).

* events for `isOpen`: `onRequestOpen` & `onRequestClose`.
* event for `selected`: `onChange`
* **behaves as outlined in the proposal**

**Comment by sh0ji** [comment]
It also occurred to me that all of our `BaseInput`-derived components have multiple controlled props:

* `errors` + `onValidate` - this is completely custom to the design system.
* `value` + `onChange` - derived from the `HTMLInputElement` interface.
* `checked` + `onChange` - derived from the `HTMLInputElement` interface.

**Comment by JRebella** [comment]
Hey @sh0ji thanks for taking the time to improve this, I think it's very important

I agree on what you describe that the component should act in one way or to other depending on the props but I think we should try to match the standard that React already has in place for `<input />` for example

The pattern I see in the props is the following:
- **value**: Should only be defined when the component will be **controlled**
- **defaultValue**: The default value of the input, should only be defined when the component will be **uncontrolled**
- **onChange**: callback that will be called every time no matter if controlled or uncontrolled. If controlled it's usually used to change the external state (value prop). Although uncommon, it can also be used in uncontrolled components to trigger side effects (for example trigger form validation once one of the uncontrolled inputs is interacted with)

So bottom line is, the **value** prop should be the "source of truth" for determining if the component will be used as controlled or uncontrolled. If it's undefined then the component is uncontrolled, if it's defined, then it's controlled

I ran some tests on https://codesandbox.io/s/objective-dream-tvy8l?file=/src/App.js 
There's some border cases and some warnings that react will throw in the console if you provide the wrong props. For example if you provide value and defaultValue at the same time, the input will be controlled and defaultValue will be ignored

There's also some border cases like when the value prop goes from defined to undefined (or vice versa) along the lifetime of the component which also causes react to throw a warning that this should not happen

**Comment by JRebella** [comment]
If we follow the React's criteria, then checkbox should be updated so that the following code renders a checked checkbox that can't be toggled:
```
<Checkbox
    checked={true}
>
    Checkbox
</Checkbox>
```

**Comment by sh0ji** [comment]
Great point. `defaultValue` and `defaultChecked` are defined by [the DOM spec for the HTMLInputElement interface](https://html.spec.whatwg.org/multipage/input.html#htmlinputelement), so we should use those for initial values and we should probably mirror that convention for custom controllable props.

We'll want to be careful about this change though, since breaking existing functionality would require a major version bump. We'll want to add this functionality while continuing to support the current functionality (even though it's inconsistent/confusing), and add deprecation warnings that recommend switching to the new functionality. Will that work for you?

**Comment by JRebella** [comment]
> 
> We'll want to be careful about this change though, since breaking existing functionality would require a major version bump. We'll want to add this functionality while continuing to support the current functionality (even though it's inconsistent/confusing), and add deprecation warnings that recommend switching to the new functionality. Will that work for you?

Yes of course! Makes total sense to me

**Comment by dmedjordan** [comment]
Issue linked to JIRA Stories: [NDS-38] [NDS-39] [NDS-40] [NDS-41] [NDS-42] [NDS-43]
